### PR TITLE
Xiangyu/to be merged to master

### PR DESCRIPTION
**Pull request by Xiangyu-Hu**
This the new update for merging.
The main changes are removing Stress_PK1_ out from ElasticSolidParticles and revised the stress and stress output.
move force_from_fluid out from particles, 
add a test case on bean pulling with load pressure,
and revision on mesh_with_data_packages,
split body relations to separate for later extensions,
move split_cell_list to real body,
and other naming and typo correction.

### CODE REVIEW

File: SPHINXsys/src/shared/particle_dynamics/solid_dynamics/contact_dynamics.cpp [hunk 1]
```diff
@@ -345,96 +336,32 @@ namespace SPH
...
-
-					// force due to pressure
-					force -= 2.0 * (impedance_p + penalty_p) * e_ij *
-							 Vol_i * Vol_k[index_j] * contact_neighborhood.dW_ij_[n];
-				}
-			}
-
-			contact_force_[index_i] = force;
-			dvel_dt_prior_[index_i] += force / mass_[index_i];
-		}
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/shared/particle_dynamics/solid_dynamics/contact_dynamics.cpp:413]: Why is it getting removed ?

**Review by FabienPean-Virtonomy — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -11,11 +11,12 @@ SPHinXsys is a multi-physics, multi-resolution SPH library.
 Although it is not a standalone application itself, 
 many examples designated for the specific type of applications are provided.
 
-[![GitHub license](https://img.shields.io/github/license/Naereen/StrapDown.js.svg)](https://github.com/Naereen/StrapDown.js/blob/master/LICENSE)
+[![GitHub license](https://img.shields.io/github/license/Naereen/StrapDown.js.svg)](https://github.com/Xiangyu-Hu/SPHinXsys/LICENSE)
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply README.md:14]: I will correct it and update the pull request.

File: SPHINXsys/src/shared/meshes/my_memory_pool.h [hunk 3]
```diff
@@ -9,64 +9,65 @@
...
-	//prepare an avaliable node
-	T* malloc()
+	// prepare an available node
+	T *malloc()
 	{
-		if (free_list.empty()) {
-			data_list.push_back(sample);
+		if (free_list.empty())
+		{
+			data_list.emplace_back(T());
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/meshes/my_memory_pool.h:46]: Ok. Let me use it.

File: SPHINXsys/src/shared/body_relations/contact_body_relation.cpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	{
+		initialization();
+	}
+	//=================================================================================================//
+	void BodyRelationContact::initialization()
+	{
+		for (size_t k = 0; k != contact_bodies_.size(); ++k)
+		{
+			CellLinkedList *target_cell_linked_list =
+				DynamicCast<CellLinkedList>(this, contact_bodies_[k]->cell_linked_list_);
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/body_relations/contact_body_relation.cpp:32]: target_cell_linked_list is temporary variable and the present dynamics cast report error if it is null and the name of class in which the cast is called.

File: SPHINXsys/src/shared/particle_dynamics/solid_dynamics/elastic_dynamics.cpp [hunk 5]
```diff
@@ -119,7 +119,7 @@ namespace SPH
 		void StressRelaxationFirstHalf::Interaction(size_t index_i, Real dt)
 		{
 			// including gravity and force from fluid
-			Vecd acceleration = acc_prior_[index_i] + force_from_fluid_[index_i] / mass_[index_i];
+			Vecd acceleration(0);
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/particle_dynamics/solid_dynamics/elastic_dynamics.cpp]: In this version all other accelerations which is collected when the velocity is updated. Therefore, we do not need add these.
See line 98 of the same file.

File: SPHINXsys/src/shared/particle_dynamics/solid_dynamics/inelastic_dynamics.cpp [hunk 6]
```diff
@@ -26,7 +26,8 @@ namespace SPH
 			F_[index_i] += dF_dt_[index_i] * dt * 0.5;
 			rho_[index_i] = rho0_ / SimTK::det(F_[index_i]);
 
-			stress_PK1_[index_i] = plastic_solid_->PlasticConstitutiveRelation(F_[index_i], index_i, dt);
+			// TODO: this formulation is temporary, to be revises.
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/particle_dynamics/solid_dynamics/inelastic_dynamics.cpp:29]: The formulation is OK. it is the name of the variable need to revise. I am sorry that my wording is misleading.

File: tests/3d_examples/test_3d_beam_pulling_pressure_load/regression_test_tool/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/regression_test_tool/BeamObserver_VonMisesStress_runtimes.dat
+        DESTINATION ${BUILD_INPUT_PATH})
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/regression_test_tool/BeamObserver_VonMisesStress_dtwdistance.xml
+        DESTINATION ${BUILD_INPUT_PATH})
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/regression_test_tool/BeamObserver_VonMisesStress_Run_0_result.xml
+        DESTINATION ${BUILD_INPUT_PATH})
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/regression_test_tool/BeamObserver_VonMisesStress_Run_3_result.xml
+        DESTINATION ${BUILD_INPUT_PATH})
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/regression_test_tool/BeamObserver_VonMisesStress_Run_5_result.xml
+        DESTINATION ${BUILD_INPUT_PATH})
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply tests/3d_examples/test_3d_beam_pulling_pressure_load/regression_test_tool/CMakeLists.txt:33]: Actually, in the later revision, I will copy the python3 file to bin, and others to a separate folder. I keep it like for now.

File: tests/3d_examples/test_3d_beam_pulling_pressure_load/regression_test_tool/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+include(ImportSPHINXsysFromSource_for_3D_build)
+
+SET(LIBRARY_OUTPUT_PATH ${PROJECT_BINARY_DIR}/lib)
+SET(EXECUTABLE_OUTPUT_PATH "${PROJECT_BINARY_DIR}/bin/")
+SET(BUILD_INPUT_PATH "${EXECUTABLE_OUTPUT_PATH}/input")
+SET(BUILD_RELOAD_PATH "${EXECUTABLE_OUTPUT_PATH}/reload")
+
+file(MAKE_DIRECTORY ${BUILD_INPUT_PATH})
+execute_process(COMMAND ${CMAKE_COMMAND} -E make_directory ${BUILD_INPUT_PATH})
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply tests/3d_examples/test_3d_beam_pulling_pressure_load/regression_test_tool/CMakeLists.txt:17]: I will revise it later.

File: tests/3d_examples/test_3d_beam_pulling_pressure_load/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/regression_test_tool/BeamObserver_VonMisesStress_runtimes.dat
+        DESTINATION ${BUILD_INPUT_PATH})
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/regression_test_tool/BeamObserver_VonMisesStress_dtwdistance.xml
+        DESTINATION ${BUILD_INPUT_PATH})
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/regression_test_tool/BeamObserver_VonMisesStress_Run_0_result.xml
+        DESTINATION ${BUILD_INPUT_PATH})
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/regression_test_tool/BeamObserver_VonMisesStress_Run_3_result.xml
+        DESTINATION ${BUILD_INPUT_PATH})
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/regression_test_tool/BeamObserver_VonMisesStress_Run_5_result.xml
+        DESTINATION ${BUILD_INPUT_PATH})
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply tests/3d_examples/test_3d_beam_pulling_pressure_load/CMakeLists.txt:33]: the same.

File: tests/3d_examples/test_3d_beam_pulling_pressure_load/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+include(ImportSPHINXsysFromSource_for_3D_build)
+
+SET(LIBRARY_OUTPUT_PATH ${PROJECT_BINARY_DIR}/lib)
+SET(EXECUTABLE_OUTPUT_PATH "${PROJECT_BINARY_DIR}/bin/")
+SET(BUILD_INPUT_PATH "${EXECUTABLE_OUTPUT_PATH}/input")
+SET(BUILD_RELOAD_PATH "${EXECUTABLE_OUTPUT_PATH}/reload")
+
+file(MAKE_DIRECTORY ${BUILD_INPUT_PATH})
+execute_process(COMMAND ${CMAKE_COMMAND} -E make_directory ${BUILD_INPUT_PATH})
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply tests/3d_examples/test_3d_beam_pulling_pressure_load/CMakeLists.txt:17]: the same.

File: SPHINXsys/src/shared/particle_dynamics/solid_dynamics/contact_dynamics.cpp (hunk 1 above)
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/particle_dynamics/solid_dynamics/contact_dynamics.cpp:413]: It was used when i suppose the ordinary contact force model is not OK dynamical cases. However, it is found it is works too. so I think this no use now.

File: SPHINXsys/src/shared/body_relations/contact_body_relation.cpp [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+				ap);
+		}
+	}
+	//=================================================================================================//
+	void SolidBodyRelationContact::initialization()
+	{
+		for (size_t k = 0; k != contact_bodies_.size(); ++k)
+		{
+			CellLinkedList *target_cell_linked_list =
+				DynamicCast<CellLinkedList>(this, contact_bodies_[k]->cell_linked_list_);
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/body_relations/contact_body_relation.cpp:98]: same answer too.

File: SPHINXsys/src/shared/body_relations/inner_body_relation.cpp [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+										 get_inner_neighbor_);
+	}
+	//=================================================================================================//
+	BodyRelationInnerVariableSmoothingLength::
+		BodyRelationInnerVariableSmoothingLength(RealBody &real_body)
+		: BaseBodyRelationInner(real_body), total_levels_(0),
+		  get_inner_neighbor_variable_smoothing_length_(&real_body)
+	{
+		MultilevelCellLinkedList *multi_level_cell_linked_list =
+			DynamicCast<MultilevelCellLinkedList>(this, real_body.cell_linked_list_);
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/body_relations/inner_body_relation.cpp:34]: The DynamicCast in SPHinXsys handles it. Please see it in ownership.h

File: SPHINXsys/src/shared/meshes/my_memory_pool.h [hunk 13]
```diff
@@ -9,64 +9,65 @@
...
-//-------------------------------------------------------------------------------------------------
-template<class T>
-class MyMemoryPool {
-	T sample;
+/**
+ * @class MyMemoryPool
+ * @brief Note that the data package T should has a default constructor.
+ */
+template <class T>
+class MyMemoryPool
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/meshes/my_memory_pool.h:17]: Currently not other place. I will consider it.

**Review by FabienPean-Virtonomy — APPROVED** [review]

**Pull request merged**